### ISSUE TITLE
Redis cache doesn't work within queue as console

### ISSUE THREAD

**Issue by timkelty** [craftcms/cms#3281]
### Description

Redis cache doesn't work within console, e.g. when running your queue as a console app, a job that tries to access the cache will fail.

### Steps to reproduce

Here's a quick and dirty way to reproduce:

1. Run your queue as console (e.g. `craft queue/listen`, with `'runQueueAutomatically' => false`)
2. Use a Redis cache, eg
    ```
        'cache' => [
            'class' => \yii\redis\Cache::class,
            'defaultDuration' => 60*60*24,
        ],
    ```
3. paste this at the top of `craft\queue\jobs\ResaveElements::execute`:
    ```
        Craft::$app->getCache()->getOrSet('foo', function() {
            return 'bar';
        });
    ```
4. Re-save a section
5. 😢 

### Resulting Error:

```
2018-09-07 16:32:33 [-][-][-][error][yii\redis\Connection] Failed to open redis DB connection (redis:6379, database=0): 0 - php_network_getaddresses: getaddrinfo failed: Name or service not known
2018-09-07 16:32:33 [-][-][-][trace][yii\db\Transaction::rollBack] Roll back transaction
2018-09-07 16:32:33 [-][-][-][error][craft\queue\QueueLogBehavior::afterError]  [1] Resaving Legal entries (attempt: 1) - Error (time: 1.429s): Failed to open redis DB connection (redis:6379, database=0): 0 - php_network_getaddresses: getaddrinfo failed: Name or service not known
2018-09-07 16:32:33 [-][-][-][error][yii\db\Exception] yii\db\Exception: Failed to open redis DB connection (redis:6379, database=0): 0 - php_network_getaddresses: getaddrinfo failed: Name or service not known in /app/vendor/composer/yiisoft/yii2-redis/src/Connection.php:565
Stack trace:
#0 /app/vendor/composer/yiisoft/yii2-redis/src/Connection.php(665): yii\redis\Connection->open()
#1 /app/vendor/composer/yiisoft/yii2-redis/src/Cache.php(199): yii\redis\Connection->executeCommand('MSET', Array)
#2 /app/vendor/composer/yiisoft/yii2/caching/Cache.php(287): yii\redis\Cache->setValues(Array, 0)
#3 /app/vendor/composer/yiisoft/yii2/caching/TagDependency.php(96): yii\caching\Cache->multiSet(Array)
#4 /app/vendor/composer/yiisoft/yii2/caching/TagDependency.php(80): yii\caching\TagDependency::touchKeys(Object(yii\redis\Cache), Array)
#5 /app/vendor/composer/nystudio107/craft-seomatic/src/services/MetaContainers.php(654): yii\caching\TagDependency::invalidate(Object(yii\redis\Cache), 'seomatic_metaco...')
#6 /app/vendor/composer/nystudio107/craft-seomatic/src/services/MetaBundles.php(409): nystudio107\seomatic\services\MetaContainers->invalidateContainerCacheByPath('', 4)
#7 /app/vendor/composer/nystudio107/craft-seomatic/src/Seomatic.php(560): nystudio107\seomatic\services\MetaBundles->invalidateMetaBundleByElement(Object(craft\elements\Entry), false)
#8 [internal function]: nystudio107\seomatic\Seomatic->nystudio107\seomatic\{closure}(Object(craft\events\ElementEvent))
#9 /app/vendor/composer/yiisoft/yii2/base/Event.php(310): call_user_func(Object(Closure), Object(craft\events\ElementEvent))
#10 /app/vendor/composer/yiisoft/yii2/base/Component.php(636): yii\base\Event::trigger('craft\\services\\...', 'afterSaveElemen...', Object(craft\events\ElementEvent))
#11 /app/vendor/composer/craftcms/cms/src/services/Elements.php(542): yii\base\Component->trigger('afterSaveElemen...', Object(craft\events\ElementEvent))
#12 /app/vendor/composer/craftcms/cms/src/services/Elements.php(1395): craft\services\Elements->saveElement(Object(craft\elements\Entry), true, false)
#13 /app/vendor/composer/craftcms/cms/src/services/Elements.php(500): craft\services\Elements->_propagateElement(Object(craft\elements\Entry), false, Array)
#14 /app/vendor/composer/craftcms/cms/src/queue/jobs/ResaveElements.php(73): craft\services\Elements->saveElement(Object(craft\elements\Entry))
#15 /app/vendor/composer/yiisoft/yii2-queue/src/Queue.php(214): craft\queue\jobs\ResaveElements->execute(Object(craft\queue\Queue))
#16 /app/vendor/composer/yiisoft/yii2-queue/src/cli/Queue.php(162): yii\queue\Queue->handleMessage('1', 'O:31:"craft\\que...', '3600', '1')
#17 /app/vendor/composer/yiisoft/yii2-queue/src/cli/Command.php(145): yii\queue\cli\Queue->execute('1', 'O:31:"craft\\que...', '3600', '1', '')
#18 [internal function]: yii\queue\cli\Command->actionExec('1', '3600', '1', '')
#19 /app/vendor/composer/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#20 /app/vendor/composer/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#21 /app/vendor/composer/yiisoft/yii2/console/Controller.php(148): yii\base\Controller->runAction('exec', Array)
#22 /app/vendor/composer/yiisoft/yii2/base/Module.php(528): yii\console\Controller->runAction('exec', Array)
#23 /app/vendor/composer/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('queue/exec', Array)
#24 /app/vendor/composer/yiisoft/yii2/console/Application.php(147): yii\console\Application->runAction('queue/exec', Array)
#25 /app/vendor/composer/yiisoft/yii2/base/Application.php(386): yii\console\Application->handleRequest(Object(craft\console\Request))
#26 /app/bin/craft(26): yii\base\Application->run()
```

### Additional info

- Craft version: 3.0.23.1
- PHP version: 7.2

Other interesting info:

- I tried access the redis cache from a simple console command directly, no problem
- With the same code pasted from the example, I don't get an error if my queue is running from web (`'runQueueAutomatically' => true`, no `queue/listen` running)
- So it seems to specifically be the combination of redis cache, within a queue job, that is running as a console app. phew.

Can anyone else that uses redis and runs their queue as console try this and see if you get the same results? /cc @boboldehampsink @ostark

**Comment by khalwat** [comment]
FWIW, I'm pretty sure this isn't an SEOmatic issue; SEOmatic just happens to leverage the Yii2 cache, whereas most of Craft does not. All I'm using is `$cache->getOrSet()` with a `TagDependency`

**Comment by brandonkelly** [comment]
The way console requests resolve host names is different than web requests, which is why sometimes `localhost` works in web, but not console.

Try changing the `REDIS_HOSTNAME` value to an IP address and see if you have better luck with that.

**Comment by timkelty** [comment]
> Try changing the REDIS_HOSTNAME value to an IP address and see if you have better luck with that.

Hmmm that's a little trickier since these are all docker containers, but I'll work something out.

Not sure if that makes total sense, though, as I'm able to connect to redis cache from a console _command_. Seems isolated to the queue-as-console, which seems weird.

**Comment by brandonkelly** [comment]
Queue-as-console is just another console command, so it should work identically to any other one.

**Comment by timkelty** [comment]
Thanks for listening everyone, figured it out.

PEBKAC on my end :)

Our `app` container didn't have network access to our `redis` one.
The joys of docker 🌈

**Comment by Natetronn** [comment]
@timkelty would you mind explaining how you hooked up network access between the two?

**Comment by timkelty** [comment]
@Natetronn just by sharing a common network via `network` (in this example, `app`):

`docker-compose.yml`:

```yml
services:
  php:
    image: php:7.3-fpm
    networks:
      - app

  redis:
    image: "redis:5.0-alpine"
    networks:
      - app
networks:
  app:
```

**Comment by Natetronn** [comment]
Thanks Tim! I actually had that part figured out before I commented here. And I figured out the other issues late last night as well. I was going to update this thread but, you beat me to it. Mayne this will help others, however. 

Anyway, being that this was on my local dev env, I had to set the hostname key from value `localhost` to `redis` instead, in the craft app.php config file setup for redis.
